# References

This file lists the scholarly works and software used in this repository. To cite this repository, see the `CITATION.cff` file.

---

## Pipeline tools

- [Snakemake](https://snakemake.github.io/)  
    >@article{koster2012snakemake, \
    >title={Snakemakeâ€”a scalable bioinformatics workflow engine}, \
    >author={K{\"o}ster, Johannes and Rahmann, Sven}, \
    >journal={Bioinformatics}, \
    >volume={28}, \
    >number={19}, \
    >pages={2520--2522}, \
    >year={2012}, \
    >publisher={Oxford University Press} \
    >}

## Software packaging/containerisation tools

- [Anaconda](https://anaconda.org/)
  > Anaconda Software Distribution. Computer software. Vers. 23.11.0. Anaconda, Dec. 2023. Web.

## Software

- [fastp](https://github.com/OpenGene/fastp)  
    >@article{chen2018fastp, \
    >title={fastp: an ultra-fast all-in-one FASTQ preprocessor}, \
    >author={Chen, Shifu and Zhou, Yanqing and Chen, Yaru and Gu, Jia}, \
    >journal={Bioinformatics}, \
    >volume={34}, \
    >number={17}, \
    >pages={i884--i890}, \
    >year={2018}, \
    >publisher={Oxford University Press} \
    >}

- [Bowtie2](http://bowtie-bio.sourceforge.net/bowtie2/index.shtml)  
    >@article{langmead2012fast, \
    >title={Fast gapped-read alignment with Bowtie 2}, \
    >author={Langmead, Ben and Salzberg, Steven L}, \
    >journal={Nature methods}, \
    >volume={9}, \
    >number={4}, \
    >pages={357--359}, \
    >year={2012}, \
    >publisher={Nature Publishing Group} \
    >}

- [samtools](https://www.htslib.org/)  
    >@article{danecek2021twelve, \
    >title={Twelve years of SAMtools and BCFtools}, \
    >author={Danecek, Petr and Bonfield, James K and Liddle, Jennifer and Marshall, John and Ohan, Valeriu and Pollard, Martin O and Whitwham, Andrew and Keane, Thomas and McCarthy, Shane A and Davies, Robert M and others}, \
    >journal={Gigascience}, \
    >volume={10}, \
    >number={2}, \
    >pages={giab008}, \
    >year={2021}, \
    >publisher={Oxford University Press} \
    >}

- [SortMeRNA](https://github.com/sortmerna/sortmerna)  
    >@article{kopylova2012sortmerna, \
    >title={SortMeRNA: fast and accurate filtering of ribosomal RNAs in metatranscriptomic data}, \
    >author={Kopylova, Evguenia and No{\'e}, Laurent and Touzet, H{\'e}l{\`e}ne}, \
    >journal={Bioinformatics}, \
    >volume={28}, \
    >number={24}, \
    >pages={3211--3217}, \
    >year={2012}, \
    >publisher={Oxford University Press} \
    >}

- [SPAdes](https://github.com/ablab/spades?tab=readme-ov-file)  
    >@article{bushmanova2019rnaspades, \
    >title={rnaSPAdes: a de novo transcriptome assembler and its application to RNA-Seq data}, \
    >author={Bushmanova, Elena and Antipov, Dmitry and Lapidus, Alla and Prjibelski, Andrey D}, \
    >journal={GigaScience}, \
    >volume={8}, \
    >number={9}, \
    >pages={giz100}, \
    >year={2019}, \
    >publisher={Oxford University Press} \
    >}

- [rnaQUAST](https://github.com/ablab/rnaquast)  
    >@article{bushmanova2016rnaquast, \
    >title={rnaQUAST: a quality assessment tool for de novo transcriptome assemblies}, \
    >author={Bushmanova, Elena and Antipov, Dmitry and Lapidus, Alla and Suvorov, Vladimir and Prjibelski, Andrey D}, \
    >journal={Bioinformatics}, \
    >volume={32}, \
    >number={14}, \
    >pages={2210--2212}, \
    >year={2016}, \
    >publisher={Oxford University Press} \
    >}

- [Kraken2](https://ccb.jhu.edu/software/kraken2/)  
    >@article{wood2019improved, \
    >title={Improved metagenomic analysis with Kraken 2}, \
    >author={Wood, Derrick E and Lu, Jennifer and Langmead, Ben}, \
    >journal={Genome biology}, \
    >volume={20}, \
    >number={1}, \
    >pages={257}, \
    >year={2019}, \
    >publisher={Springer} \
    >}

- [Bracken](https://ccb.jhu.edu/software/bracken/)  
    >@article{lu2017bracken, \
    >title={Bracken: estimating species abundance in metagenomics data}, \
    >author={Lu, Jennifer and Breitwieser, Florian P and Thielen, Peter and Salzberg, Steven L}, \
    >journal={PeerJ Computer Science}, \
    >volume={3}, \
    >pages={e104}, \
    >year={2017}, \
    >publisher={PeerJ Inc.} \
    >}

- [RGI](https://github.com/arpcard/rgi)  
    >@article{alcock2023card, \
    >title={CARD 2023: expanded curation, support for machine learning, and resistome prediction at the Comprehensive Antibiotic Resistance Database}, \
    >author={Alcock, Brian P and Huynh, William and Chalil, Romeo and Smith, Keaton W and Raphenya, Amogelang R and Wlodarski, Mateusz A and Edalatmand, Arman and Petkau, Aaron and Syed, Sohaib A and Tsang, Kara K and others}, \
    >journal={Nucleic acids research}, \
    >volume={51}, \
    >number={D1}, \
    >pages={D690--D699}, \
    >year={2023}, \
    >publisher={Oxford University Press} \
    >}

- [MEGAHIT](https://github.com/voutcn/megahit)  
    >@article{li2015megahit, \
    >title={MEGAHIT: an ultra-fast single-node solution for large and complex metagenomics assembly via succinct de Bruijn graph}, \
    >author={Li, Dinghua and Liu, Chi-Man and Luo, Ruibang and Sadakane, Kunihiko and Lam, Tak-Wah}, \
    >journal={Bioinformatics}, \
    >volume={31}, \
    >number={10}, \
    >pages={1674--1676}, \
    >year={2015}, \
    >publisher={Oxford University Press} \
    >}

- [Prodigal](https://github.com/hyattpd/Prodigal)  
    >@article{hyatt2010prodigal, \
    >title={Prodigal: prokaryotic gene recognition and translation initiation site identification}, \
    >author={Hyatt, Doug and Chen, Gwo-Liang and LoCascio, Philip F and Land, Miriam L and Larimer, Frank W and Hauser, Loren J}, \
    >journal={BMC bioinformatics}, \
    >volume={11}, \
    >number={1}, \
    >pages={119}, \
    >year={2010}, \
    >publisher={Springer} \
    >}

- [featureCounts](https://subread.sourceforge.net/featureCounts.html)  
    >@article{liao2014featurecounts, \
    >title={featureCounts: an efficient general purpose program for assigning sequence reads to genomic features}, \
    >author={Liao, Yang and Smyth, Gordon K and Shi, Wei}, \
    >journal={Bioinformatics}, \
    >volume={30}, \
    >number={7}, \
    >pages={923--930}, \
    >year={2014}, \
    >publisher={Oxford University Press} \
    >}

### Databases

- [CARD](https://card.mcmaster.ca/)  
    >@article{alcock2023card, \
    >title={CARD 2023: expanded curation, support for machine learning, and resistome prediction at the Comprehensive Antibiotic Resistance Database}, \
    >author={Alcock, Brian P and Huynh, William and Chalil, Romeo and Smith, Keaton W and Raphenya, Amogelang R and Wlodarski, Mateusz A and Edalatmand, Arman and Petkau, Aaron and Syed, Sohaib A and Tsang, Kara K and others}, \
    >journal={Nucleic acids research}, \
    >volume={51}, \
    >number={D1}, \
    >pages={D690--D699}, \
    >year={2023}, \
    >publisher={Oxford University Press} \
    >}

- [GTDB](https://gtdb.ecogenomic.org/about)  
    >@article{parks2022gtdb, \
    >title={GTDB: an ongoing census of bacterial and archaeal diversity through a phylogenetically consistent, rank normalized and complete genome-based taxonomy}, \
    >author={Parks, Donovan H and Chuvochina, Maria and Rinke, Christian and Mussig, Aaron J and Chaumeil, Pierre-Alain and Hugenholtz, Philip}, \
    >journal={Nucleic acids research}, \
    >volume={50}, \
    >number={D1}, \
    >pages={D785--D794}, \
    >year={2022}, \
    >publisher={Oxford University Press} \
    >}

- [BUSCO](https://busco.ezlab.org/)  
    >@article{tegenfeldt2025orthodb, \
    >title={OrthoDB and BUSCO update: annotation of orthologs with wider sampling of genomes}, \
    >author={Tegenfeldt, Fredrik and Kuznetsov, Dmitry and Manni, Mos{\`e} and Berkeley, Matthew and Zdobnov, Evgeny M and Kriventseva, Evgenia V}, \
    >journal={Nucleic Acids Research}, \
    >volume={53}, \
    >number={D1}, \
    >pages={D516--D522}, \
    >year={2025}, \
    >publisher={Oxford University Press} \
    >}
